File: backend/ml_model/train_model.js
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import train_test_split, RandomizedSearchCV, cross_val_score
from imblearn.over_sampling import SMOTE

# Load the dataset
data = pd.read_csv('preprocessed_health_data.csv')

# Display basic information
print("Dataset Info:")
print(data.info())

# Check for missing values
print("Missing Values in Each Column:")
print(data.isnull().sum())

# Check target variable distribution
print("Target Variable Distribution:")
print(data['disease'].value_counts())

# Select features and target variable
X = data.drop('disease', axis=1)
y = data['disease']

# Scale the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Define and train the RandomForest model
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_scaled, y)

# Hyperparameter Tuning
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}
random_search = RandomizedSearchCV(estimator=rf_model, param_distributions=param_grid,
                                   n_iter=10, cv=3, verbose=2, random_state=42, n_jobs=-1)
random_search.fit(X_scaled, y)
best_params = random_search.best_params_
print("Best parameters found: {}".format(best_params))
best_rf_model = random_search.best_estimator_

# Feature Importance
importances = best_rf_model.feature_importances_
feature_names = X.columns
feature_importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': importances})
important_features = feature_importance_df.sort_values(by='Importance', ascending=False)
print("Feature Importance:")
print(important_features.head(20))

# Handle Class Imbalance with SMOTE
smote = SMOTE()
X_resampled, y_resampled = smote.fit_resample(X_scaled, y)
best_rf_model.fit(X_resampled, y_resampled)

# Cross-Validation
cv_scores_resampled = cross_val_score(best_rf_model, X_resampled, y_resampled, cv=5, scoring='accuracy')
print("Cross-Validation Accuracy Scores on Resampled Data: {}".format(cv_scores_resampled))
print("Mean Accuracy on Resampled Data: {}".format(cv_scores_resampled.mean()))

# Final Evaluation
y_pred = best_rf_model.predict(X_scaled)
print("Confusion Matrix:")
print(confusion_matrix(y, y_pred))
print("Classification Report:")
print(classification_report(y, y_pred))

File: backend/ml_model/train_model.py
# backend\ml_model\train_model.py

import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns
import os

# Set random seed for reproducibility
tf.random.set_seed(42)
np.random.seed(42)

# Load the dataset
data_path = 'preprocessed_health_data.csv'  # Update this path if necessary
data = pd.read_csv(data_path)

# Inspect the data
print("Dataset Shape:", data.shape)
print("Column Names:", data.columns)
print("First 5 Rows:\n", data.head())

# Separate features and target
X = data.drop(columns=['disease'])
y = data['disease']

# Encode target labels
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)
num_classes = len(label_encoder.classes_)
print("Classes:", label_encoder.classes_)

# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(
    X, y_encoded, test_size=0.2, random_state=42
)

# Build the model
model = Sequential([
    Dense(256, activation='relu', input_shape=(X_train.shape[1],)),
    Dropout(0.5),
    Dense(128, activation='relu'),
    Dropout(0.3),
    Dense(64, activation='relu'),
    Dense(num_classes, activation='softmax')
])

# Compile the model
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# Train the model
history = model.fit(
    X_train, y_train,
    epochs=50,
    batch_size=32,
    validation_data=(X_test, y_test)
)

# Evaluate the model
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"\nTest Accuracy: {test_accuracy*100:.2f}%")

# Generate classification report
y_pred = model.predict(X_test)
y_pred_classes = np.argmax(y_pred, axis=1)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_classes, target_names=label_encoder.classes_))

# Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred_classes)
plt.figure(figsize=(12, 10))
sns.heatmap(conf_matrix, annot=True, fmt='d', xticklabels=label_encoder.classes_, yticklabels=label_encoder.classes_, cmap='Blues')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.title('Confusion Matrix')
plt.tight_layout()
plt.savefig('confusion_matrix.png')
plt.close()

# Save the model and label encoder
model.save('disease_prediction_model.h5')
with open('label_encoder.npy', 'wb') as f:
    np.save(f, label_encoder.classes_)

print("\nModel and label encoder saved successfully.")
